modulus/distributed/config.py

```diff
@@ -0,0 +1,259 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from treelib import Tree
+from typing import Dict, List, Optional, Union
+
+
+class ProcessGroupNode:
+    """
+    Class to store the attributes of a distributed process group
+
+    Attributes
+    ----------
+    name : str
+        Name of the process group
+    size : Optional[int]
+        Optional, size of the process group
+    orthogonal_group : Optional[str]
+        Optional, name of an orthogonal process group to create
+    """
+
+    def __init__(
+        self,
+        name: str,
+        size: Optional[int] = None,
+        orthogonal_group: Optional[str] = None,
+    ):
+        """
+        Constructor for the ProcessGroupNode class
+
+        Parameters
+        ----------
+        name : str
+            Name of the process group
+        size : Optional[int]
+            Optional, size of the process group
+        orthogonal_group : Optional[str]
+            Optional, name of an orthogonal process group to create
+        """
+        self.name = name
+        self.size = size
+        self.orthogonal_group = orthogonal_group
+
+    def __str__(self):
+        """
+        String representation of the process group node
+
+        Returns
+        -------
+        str
+            String representation of the process group node
+        """
+        return (
+            "ProcessGroupNode("
+            f"name={self.name}, "
+            f"size={self.size}, "
+            f"orthogonal_group={self.orthogonal_group})"
+        )
+
+    def __repr__(self):
+        """
+        String representation of the process group node
+
+        Returns
+        -------
+        str
+            String representation of the process group node
+        """
+        return self.__str__()
+
+
+class ProcessGroupConfig:
+    """
+    Class to define the configuration of a model's parallel process group structure as a
+    tree. Each node of the tree is of type `ProcessGroupNode`.
+
+    Once the process group config structure (i.e, the tree structure) is set, it is
+    sufficient to set only the sizes for each leaf process group. Then, the size of
+    every parent group can be automatically computed as the product reduction of the
+    sub-tree of that parent group node.
+
+    Examples
+    --------
+    >>> from modulus.distributed.config import ProcessGroupNode, ProcessGroupConfig
+    >>> mp = ProcessGroupNode("model_parallel", orthogonal_group="data_parallel")
+    >>>
+    >>> # Create model parallel group with data parallel as the orthogonal group
+    >>> mp = ProcessGroupNode("model_parallel", orthogonal_group="data_parallel")
+    >>>
+    >>> # Create the process group config with the highest level process group
+    >>> config = ProcessGroupConfig(mp)
+    >>>
+    >>> # Create spatial and channel parallel sub-groups
+    >>> config.add_node(ProcessGroupNode("spatial_parallel"), parent=mp)
```

Comment:
Can we add nodes to the `orthogonal_group`?

Reply:
In the current design, no. Is there a case where this might be needed? We can structure the tree as a binary tree with the right node always being the orthogonal of the left but that has less than ideal usability and is also invariant under certain permutations making it a bit harder to deal with from both a user and library perspective.

Reply:
@dallasfoster https://github.com/NVIDIA/modulus/pull/188/files#diff-953c2b18aa629b13fc90a0c3df4fd3adfcbcf7407651517ce6844668f430a5f7R524 might help explain how we might reformulate this process group config as a binary tree with the right node always being the orthogonal of the corresponding left node.